ui/decredmaterial/editor.go

```diff
@@ -173,27 +147,78 @@ func (e Editor) Layout(gtx layout.Context) layout.Dimensions {
 	})
 }
 
-func (e Editor) editorLine(gtx C) D {
+func (e Editor) editorLayout(gtx C) D {
+	var dims layout.Dimensions
+	switch e.LineStyle {
+	case RoundedRectangle:
+		dims = e.editorRectangle(gtx, func(gtx C) D {
+			return e.editorSection(gtx, false)
+		})
+	case SingleUnderLine:
+		dims = e.editorSection(gtx, true)
+	case NoLine:
+		dims = e.editorSection(gtx, false)
+	}
+	return dims
+}
+
+func (e Editor) editorSection(gtx layout.Context, underline bool) layout.Dimensions {
 	return layout.Flex{}.Layout(gtx,
+		layout.Flexed(1, func(gtx C) D {
+			return layout.Flex{Axis: layout.Vertical}.Layout(gtx,
+				layout.Rigid(func(gtx C) D {
+					m := values.MarginPadding5
+					inset := layout.Inset{
+						Top:    m,
+						Bottom: m,
+					}
+					return inset.Layout(gtx, func(gtx C) D {
+						return e.EditorStyle.Layout(gtx)
+					})
+				}),
+				layout.Rigid(func(gtx C) D {
+					if underline {
+						return e.editorLine(gtx)
+					}
+					return layout.Dimensions{}
+				}),
+			)
+		}),
 		layout.Rigid(func(gtx C) D {
-			dims := image.Point{
-				X: gtx.Constraints.Max.X,
-				Y: 2,
-			}
-			rect := f32.Rectangle{
-				Max: layout.FPt(dims),
+			if e.IsVisible {
+				inset := layout.Inset{
+					Top:  values.MarginPadding2,
+					Left: values.MarginPadding5,
+				}
+				return inset.Layout(gtx, func(gtx C) D {
+					if e.Editor.Text() == "" {
+						return e.pasteBtnMaterial.Layout(gtx)
+					}
+					return e.clearBtMaterial.Layout(gtx)
+				})
 			}
-			op.Offset(f32.Point{
-				X: 0,
-				Y: 0,
-			}).Add(gtx.Ops)
-			paint.ColorOp{Color: e.LineColor}.Add(gtx.Ops)
-			paint.PaintOp{Rect: rect}.Add(gtx.Ops)
-			return layout.Dimensions{Size: dims}
+			return layout.Dimensions{}
 		}),
 	)
 }
 
+func (e Editor) editorRectangle(gtx layout.Context, body layout.Widget) layout.Dimensions {
+	border := widget.Border{Color: e.LineColor, CornerRadius: values.MarginPadding5, Width: values.MarginPadding1}
+	return border.Layout(gtx, func(gtx C) D {
+		mtb := values.MarginPadding2
+		mlr := values.MarginPadding5
```

Comment:
you can define the margins and paddings within the widgets. decredmaterial widgets should be independent of variables and definitions outside the decredmaterial package.

Reply:
So basically we cannot use the dimensions values declared already.

Reply:
not in decredmaterial